Branch-SDK/Branch-SDK/BNCAppleSearchAds.m

```diff
@@ -0,0 +1,220 @@
+@property (nonatomic, assign, readwrite) SEL adClientRequestAttribution;
+@end
+
+@implementation BNCAppleSearchAds
+
++ (BNCAppleSearchAds *)sharedInstance {
+    static BNCAppleSearchAds *singleton;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        singleton = [[BNCAppleSearchAds alloc] init];
+    });
+    return singleton;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.enableAppleSearchAdsCheck = NO;
+        self.adClientClass = NSClassFromString(@"ADClient");
+        self.adClientSharedClient = NSSelectorFromString(@"sharedClient");
+        self.adClientRequestAttribution = NSSelectorFromString(@"requestAttributionDetailsWithBlock:");
+    }
+    return self;
+}
+
+// business logic around checking and storing Apple Search Ads attribution
+- (void)checkAppleSearchAdsSaveTo:(BNCPreferenceHelper *)preferenceHelper installDate:(NSDate *)installDate completion:(void (^_Nullable)(void))completion {
+    if (!self.enableAppleSearchAdsCheck) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if ([self isAppleSearchAdSavedToDictionary:preferenceHelper.appleSearchAdDetails]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if (![self isDateWithinWindow:installDate]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    if (![self isAdClientAvailable]) {
+        if (completion) {
+            completion();
+        }
+        return;
+    }
+    
+    [self requestAttributionWithCompletion:^(NSDictionary * _Nullable attributionDetails, NSError * _Nullable error, NSTimeInterval elapsedSeconds) {
+        // BNCPreferenceHelper should be responsible for correctly storing and resetting this
+        @synchronized (preferenceHelper) {
+            if (attributionDetails.count > 0 && !error) {
+                [preferenceHelper addInstrumentationDictionaryKey:@"apple_search_ad" value:[[NSNumber numberWithInteger:elapsedSeconds*1000] stringValue]];
+            }
+            if (!error) {
+                if (attributionDetails == nil) {
+                    attributionDetails = @{};
+                }
+                if (preferenceHelper.appleSearchAdDetails == nil) {
+                    preferenceHelper.appleSearchAdDetails = @{};
+                }
+                if (![preferenceHelper.appleSearchAdDetails isEqualToDictionary:attributionDetails]) {
+                    preferenceHelper.appleSearchAdDetails = attributionDetails;
+                    preferenceHelper.appleSearchAdNeedsSend = YES;
+                }
+            }
+        }
+        
+        if (completion) {
+            completion();
+        }
+    }];
+}
+
+- (BOOL)isAppleSearchAdSavedToDictionary:(NSDictionary *)appleSearchAdDetails {
+    NSDictionary *tmp = [appleSearchAdDetails objectForKey:@"Version3.1"];
```

Comment:
This is a non-actionable comment but I'm leaving it here, so we can chat about it.. a key that's a version number sounds crazy to me..

Reply:
That's Apple's json format.  :(